src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs

```diff
@@ -69,31 +70,73 @@ private async Task<string> GetRegionAsync(ICoreLogger logger)
+                {
+                    { "Metadata", "true" }
+                };
+                
+                HttpResponse response = await _httpManager.SendGetAsync(BuildImdsUri(DefaultApiVersion), headers, logger).ConfigureAwait(false);
+
+                // A bad request occurs when the version in the IMDS call is no longer supported.
+                if (response.StatusCode == HttpStatusCode.BadRequest)
+                {
+                    string apiVersion = await GetImdsUriApiVersionAsync(logger, headers).ConfigureAwait(false); // Get the latest version
+                    response = await _httpManager.SendGetAsync(BuildImdsUri(apiVersion), headers, logger).ConfigureAwait(false); // Call again with updated version
+                }
 
                 if (response.StatusCode != HttpStatusCode.OK)
                 {
-                    throw new MsalClientException(
-                        MsalError.RegionDiscoveryFailed,
-                        MsalErrorMessage.RegionDiscoveryFailed);
+                    throw MsalServiceExceptionFactory.FromGeneralHttpResponse(
+                    MsalError.RegionDiscoveryFailed,
+                    MsalErrorMessage.RegionDiscoveryFailed,
+                    response);
                 }
 
                 LocalImdsResponse localImdsResponse = JsonHelper.DeserializeFromJson<LocalImdsResponse>(response.Body);
 
                 logger.Info($"[Region discovery] Call to local IMDS returned region: {localImdsResponse.location}");
                 return localImdsResponse.location;
             }
-            catch (MsalClientException)
+            catch (MsalServiceException)
             {
                 throw;
             }
             catch (Exception e)
             {
                 logger.Info("[Region discovery] Call to local imds failed." + e.Message);
-                throw new MsalClientException(MsalError.RegionDiscoveryFailed, MsalErrorMessage.RegionDiscoveryFailed);
+                throw new MsalServiceException(MsalError.RegionDiscoveryFailed, MsalErrorMessage.RegionDiscoveryFailed);
             }
         }
 
+        private async Task<string> GetImdsUriApiVersionAsync(ICoreLogger logger, Dictionary<string, string> headers)
+        {
+            Uri imdsErrorUri = new Uri(ImdsEndpoint);
+
+            HttpResponse response = await _httpManager.SendGetAsync(imdsErrorUri, headers, logger).ConfigureAwait(false);
+
+            // When IMDS endpoint is called without the api version query param, bad request response comes back with latest version.
+            if (response.StatusCode == HttpStatusCode.BadRequest)
+            {
+                LocalImdsErrorResponse errorResponse = JsonHelper.DeserializeFromJson<LocalImdsErrorResponse>(response.Body);
+                logger.Info("[Region discovery] Updated the version for IMDS endpoint to: " + errorResponse.NewestVersions[0]);
+                return errorResponse.NewestVersions[0];
```

Comment:
Not defensive enough. What happens if the error response does not contain the newest versions - you'll get a null reference exception which is not easy to debug. 

I would suggest throwing a service exception similar to the one at line 127 (same error code, maybe slightly different error message).